dfdaemon/proxy/proxy.go

```diff
@@ -314,7 +322,39 @@ func (proxy *Proxy) handleHTTPS(w http.ResponseWriter, r *http.Request) {
 	logrus.Debugln("hijack https request to", r.Host)
 
 	sConfig := new(tls.Config)
-	sConfig.Certificates = []tls.Certificate{*proxy.cert}
+	if proxy.cert.Leaf != nil && proxy.cert.Leaf.IsCA {
+		if proxy.certCache == nil { // Initialize proxy.certCache on first access. (Lazy init)
+			proxy.certCache = lru.New(100) // Default max entries size = 100
+		}
+		logrus.Debugf("hijack https request with CA <%s>", proxy.cert.Leaf.Subject.CommonName)
+		leafCertSpec := LeafCertSpec{
+			proxy.cert.Leaf.PublicKey,
+			proxy.cert.PrivateKey,
+			proxy.cert.Leaf.SignatureAlgorithm}
+		host, _, _ := net.SplitHostPort(r.Host)
+		sConfig.GetCertificate = func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			cConfig.ServerName = hello.ServerName
+			logrus.Debugf("Generate temporal leaf TLS cert for ServerName <%s>, host <%s>", hello.ServerName, host)
+			// cacheKey is (ServerName, host) pair. Just use string concatenation, no need for a tuple struct here.
+			// I'm curious about whether `hello.ServerName` is always same as `host`?
```

Comment:
In theory `hello.ServerName` is the CN of the certificate, which according to the comment in the std library and [RFC4366 section 3.1](https://tools.ietf.org/html/rfc4366#section-3.1) may be different from `host`. 

Though come to think of it, there might actually be a special case. A certificate can be used on an ip address, for example, cloudflare's [1.1.1.1](https://1.1.1.1/). In such case, `hello.ServerName` would be empty, and adding `host`, which is an ip address, to `Certificate.DNSNames` in `genLeafCert`, will result in an invalid certificate. 

So I think we can ignore `hello.ServerName`, and in `genLeafCert`, we'll use `host` as CN, and add `host` to `Certificate.IPAddresses` if it's an ip, otherwise add it to `Certificate.DNSNames`.

Reply:
You mean I need to add another check to determine whether the `host` is IP or domain, and add it to `IPAddresses` and `DNSNames` accordingly?

What about the `cacheKey`? Just use `host` value without associating it with `hello.ServerName` ?

Reply:
> You mean I need to add another check to determine whether the `host` is IP or domain, and add it to `IPAddresses` and `DNSNames` accordingly?

Yes, so it will work for both domain name and ip address.

> What about the `cacheKey`? Just use `host` value without associating it with `hello.ServerName` ?

Yes, since we're ignoring `hello.ServerName`, `host` alone will suffice. And we can also change the parameters of `genLeafCert`:

```
func genLeafCert(ca *tls.Certificate, leafCertSpec *LeafCertSpec, host string) (*tls.Certificate, error)
```

Reply:
Understood. I updated the PR.